src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,141 @@
+    /// Computes the semantic tokens edits for a file. An edit request is received every 500ms,
+    /// or every time an edit is made by the user.
+    /// </summary>
+    [ExportLspMethod(LSP.SemanticTokensMethods.TextDocumentSemanticTokensEditsName), Shared]
+    internal class SemanticTokensEditsHandler : AbstractRequestHandler<LSP.SemanticTokensEditsParams, SumType<LSP.SemanticTokens, LSP.SemanticTokensEdits>>
+    {
+        private readonly SemanticTokensCache _tokensCache;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SemanticTokensEditsHandler(
+            ILspSolutionProvider solutionProvider,
+            SemanticTokensCache tokensCache) : base(solutionProvider)
+        {
+            _tokensCache = tokensCache;
+        }
+
+        public override async Task<SumType<LSP.SemanticTokens, LSP.SemanticTokensEdits>> HandleRequestAsync(
+            LSP.SemanticTokensEditsParams request,
+            LSP.ClientCapabilities clientCapabilities,
+            string? clientName,
+            CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfNull(request.TextDocument);
+            var resultId = await _tokensCache.GetNextResultIdAsync(request.TextDocument.Uri, cancellationToken).ConfigureAwait(false);
+
+            // Even though we want to ultimately pass edits back to LSP, we still need to compute all semantic tokens,
+            // both for caching purposes and in order to have a baseline comparison when computing the edits.
+            var updatedTokens = await SemanticTokensHelpers.ComputeSemanticTokensAsync(
+                request.TextDocument, resultId, clientName, SolutionProvider,
+                range: null, cancellationToken).ConfigureAwait(false);
+
+            // Getting the cached tokens for the document. If we don't have an applicable cached token set,
+            // we can't calculate edits, so we must return all semantic tokens instead.
+            var cachedTokens = await _tokensCache.GetCachedTokensAsync(
+                request.TextDocument.Uri, request.PreviousResultId, cancellationToken).ConfigureAwait(false);
+            if (cachedTokens == null)
+            {
+                return updatedTokens;
+            }
+
+            // The Data property is always populated on the server side, so it should never be null.
+            Contract.ThrowIfNull(cachedTokens.Data);
+            Contract.ThrowIfNull(updatedTokens.Data);
+
+            var edits = ComputeSemanticTokensEdits(resultId, cachedTokens.Data, updatedTokens.Data);
+            await _tokensCache.UpdateCacheAsync(request.TextDocument.Uri, updatedTokens, cancellationToken).ConfigureAwait(false);
+            return edits;
+        }
+
+        /// <summary>
+        /// Compares two sets of SemanticTokens and returns the edits between them.
+        /// </summary>
+        private static LSP.SemanticTokensEdits ComputeSemanticTokensEdits(
```

Comment:
Does the size of edits matter? If I'm reading this right, for an insertion in the middle of a file this algorithm will produce two edits, rather than one? And the first of the two will be potentially very large.